Singlenozzle Switching Extruder (Servo move on tool-change)

schimmi70 opened the pull request:
With this change it is possible to use up to 4 filaments at the same time with only two stepper drivers on the mainboard of printer.

schimmi70 committed:
Changes for use 2 swichting extruder / stepper with only one Hotend

Now it is possible to use up for 4 filaments at the same time with only two stepper-driver on the mainboard of printer

thinkyhead commented:
Compile errors and conflicts with the latest code. Please `git rebase` (or `git merge`) this with the latest code in `RCBugFix` and repair the compile error(s) [reported by Travis CI](https://travis-ci.org/MarlinFirmware/Marlin/builds/219276958), then update the PR.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -49,7 +49,7 @@
  * Note: Update also Version.h !
  */
-#define CONFIGURATION_H_VERSION 010100
+#define CONFIGURATION_H_VERSION 010101
```
Don't include all these changes in `Configuration.h`. Only the new feature should be added.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -156,8 +156,20 @@
+
+// A single extruder that uses one or two single stepper motor
+// Don't forget to set SSDE_SERVO_ANGLES 
+#define SINGLENOZZLE_SWITCHING_EXTRUDER
```
This file is the base/demo configuration for all users. So, you should include only these changes, adding the new extruder type. And `SINGLENOZZLE_SWITCHING_EXTRUDER` must be disabled by default.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5404,7 +5404,7 @@ inline void gcode_M42() {
    *     S = Schizoid (Or Star if you prefer)
    *
-   * This function assumes the bed has been homed.  Specifically, that a G28 command
+   * This function assumes the bed has been homed.  Specifically, that a  command
```
This edit is not needed. Please undo.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5688,11 +5688,18 @@ inline void gcode_M77() { print_job_timer.stop(); }
 /**
  * M104: Set hot end temperature
  */
+#if ENABLED(SINGLENOZZLE) || ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
```
Indentation of `#if` here should be 2 spaces, with contents indented 2 more spaces.

schimmi70 replied:
i don't know, that i understand you right and i hope i changed it correct.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5885,7 +5892,10 @@ inline void gcode_M109() {
   if (DEBUGGING(DRYRUN)) return;
 
-  #if ENABLED(SINGLENOZZLE)
+ #if ENABLED(SINGLENOZZLE) || ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
```
Please preserve the indentation cascade.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8466,6 +8476,22 @@ inline void gcode_M999() {
+  inline void move_extruder_servo(uint8_t e) {
+  #if EXTRUDERS == 2
+      const int angles_e1_e2[2] = SWITCHING_EXTRUDER_SERVO_E1_E2_ANGLES;
+      MOVE_SERVO(SWITCHING_EXTRUDER_SERVO_E1_E2_NR, angles_e1_e2[e]);
```
Only 2 spaces indentation on these lines.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -321,16 +321,21 @@
   #endif
 
-#elif ENABLED(SINGLENOZZLE)
+#elif ENABLED(SINGLENOZZLE) 
```
Careful about trailing white-space. A regex search on `[ \t]+$` will locate all instances of trailing whitespace.

thinkyhead commented on Marlin/configuration_store.cpp:
```diff
@@ -1098,7 +1098,7 @@ void Config_ResetDefault() {
     SERIAL_EOL;
     #if ENABLED(DISTINCT_E_FACTORS)
-      for (uint8_t i = 0; i < E_STEPPERS; i++) {
+      for (uint8_t i = 0; i < EXTRUDERS; i++) {
```
These should be left as `E_STEPPERS`.

Here we have to be cautious. The `axis_steps_per_mm` array length (`XYZE_N`) is based on `E_STEPPERS`, and these speed and acceleration factors do pertain to the steppers, not to the "extruders." The `DISTINCT_E_FACTORS` configuration option is what determines whether all steppers share the same steps-per-unit (etc.) or whether they each have their own distinct value. When they do have their own values, the `active_extruder` is used to index into these arrays.

schimmi70 replied:
There is a simple problem... in the case of switching extruder we have only one or to stepper for two or four "pseudo" extruder so if you change to extruder higher E_STEPPERS DISTINCT_E_FACTORS don't work. 
I think, if i made it dependet from the SWITCHING_EXTRUDER parameter it can be OK !?

schimmi70 replied:
is there another case, when except switching extruder Number of E_STEPPERS is not equal to number of EXTRUDERS ? 

I think MIXING-Extruder, but do they use EXTRUDER Parameter  ?

thinkyhead commented on Marlin/pins.h:
```diff
@@ -308,7 +308,12 @@
 #define _E4_PINS
 
-#if EXTRUDERS > 1
+#if ENABLED (SINGLENOZZLE_SWITCHING_EXTRUDER) // needs only one stepper for 2 Extruder
```
No space between `ENABLED` and the opening parenthesis.

thinkyhead commented on Marlin/pins.h:
```diff
@@ -308,7 +308,12 @@
 
-#if EXTRUDERS > 1
+#if ENABLED (SINGLENOZZLE_SWITCHING_EXTRUDER) // needs only one stepper for 2 Extruder
+  #if EXTRUDERS > 2
```
Since `SINGLENOZZLE_SWITCHING_EXTRUDER` uses exactly 2 or 4 extruders, then this block should actually be:
```cpp
#if ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
  #undef _E1_PINS
  #define _E1_PINS E1_STEP_PIN, E1_DIR_PIN, E1_ENABLE_PIN, E1_MS1_PIN, E1_MS2_PIN,
  #if EXTRUDERS == 4
    #undef _E2_PINS
    #define _E2_PINS E2_STEP_PIN, E2_DIR_PIN, E2_ENABLE_PIN, E2_MS1_PIN, E2_MS2_PIN,
    #undef _E3_PINS
    #define _E3_PINS E3_STEP_PIN, E3_DIR_PIN, E3_ENABLE_PIN, E3_MS1_PIN, E3_MS2_PIN,
  #endif
#elif EXTRUDERS > 1
  . . .
#endif
```

schimmi70 replied:
E2 / E3 pins are not used, cause only max two stepper E0 and E1 ...

In this section a E4 Stepper (Extruder 5 !? )  is defined, i think it shoud be only max E3 !?

thinkyhead commented on Marlin/pins_ULTIMAKER.h:
```diff
@@ -37,23 +37,24 @@
+#define Y_MAX_PIN          -1 // 28
 #define Z_MIN_PIN          30
-#define Z_MAX_PIN          32
+#define Z_MAX_PIN          -1 // 32
```
Don't include these changes to the ultimaker pins file.

thinkyhead commented on Marlin/pins_ULTIMAKER.h:
```diff
@@ -79,21 +80,33 @@
+#define E3_STEP_PIN        E2_STEP_PIN
+#define E3_DIR_PIN         E2_DIR_PIN
+#define E3_ENABLE_PIN      E2_ENABLE_PIN
+
```
I'm not sure of the overall effect of having E1, E2, and E3 all point to the same pins, all the time. Please hold off on this change. If this is important to making `SINGLENOZZLE_SWITCHING_EXTRUDER` work, then we need to find a more universal method that works for all pins files. (I'm sure we shouldn't go into all the pins files and add duplicated E pins, as here.)

thinkyhead commented on Marlin/pins_ULTIMAKER.h:
```diff
@@ -79,21 +80,33 @@
-#define TEMP_1_PIN          9   // Analog Input
+#define TEMP_1_PIN          8   // 9   // Analog Input
+#define TEMP_2_PIN          8   // Analog Input
+#define TEMP_3_PIN          8   // Analog Input
```
Again, we should not include these customizations.

thinkyhead commented on Marlin/pins_ULTIMAKER.h:
```diff
@@ -79,21 +80,33 @@
 #define HEATER_BED_PIN      4
 
-#define FAN_PIN             7
+#define FAN_PIN             7 
```
Excess whitespace.

thinkyhead commented on Marlin/stepper_indirection.h:
```diff
@@ -388,6 +388,16 @@
+  #elif EXTRUDERS == 4
+    #define E_STEP_WRITE(v) { switch (current_block->active_extruder) { case 0: E0_STEP_WRITE(v); break; case 1: E0_STEP_WRITE(v); break; case 2: E1_STEP_WRITE(v); break; case 3: E1_STEP_WRITE(v); } }
+    #define NORM_E_DIR() { switch (current_block->active_extruder) { case 0: E0_DIR_WRITE(!INVERT_E0_DIR); break; case 1: E0_DIR_WRITE(INVERT_E0_DIR); break; case 2: E1_DIR_WRITE(!INVERT_E1_DIR); break; case 3: E1_DIR_WRITE(INVERT_E1_DIR); } }
+    #define REV_E_DIR() { switch (current_block->active_extruder) { case 0: E0_DIR_WRITE(INVERT_E0_DIR); break; case 1: E0_DIR_WRITE(!INVERT_E0_DIR); break; case 2: E1_DIR_WRITE(INVERT_E1_DIR); break; case 3: E1_DIR_WRITE(!INVERT_E1_DIR); } }
```
Since the `EXTRUDERS == 4` version only uses steppers E0 and E1 anyway, it does appear that the extra pins in the Ultimaker file aren't needed, after all.

Anyway, here's a slightly shorter first line:
```cpp
#define E_STEP_WRITE(v) { if (current_block->active_extruder < 2) E0_STEP_WRITE(v); else E1_STEP_WRITE(v); }
```

schimmi70 replied:
That is right, most modification in the pins_ULTIMAKER.h are form early test to make the switching extruder work. 
And i forgot to delete them. Now only 2 E_Stepper / 1 temp / 1 hotend and the 2 servo pins an nessary.
i corecct ist and test it one more time. bevor i push the changes.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -454,7 +454,11 @@ void Temperature::updatePID() {
+    return heater < 0 ? soft_pwm_bed : soft_pwm[0];
+  #else
+    return heater < 0 ? soft_pwm_bed : soft_pwm[heater];
+  #endif
```
To keep repetition to a minimum, this form is preferred:
```cpp
return heater < 0 ? soft_pwm_bed : soft_pwm[
  #if HOTENDS == 1
    0
  #else
    heater
  #endif
];
```

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -63,7 +63,7 @@ char lcd_status_message[3 * (LCD_WIDTH) + 1] = WELCOME_MSG; // worst case is kan
   #include "ultralcd_impl_HD44780.h"
 #endif
-
+ 
```
Excess whitespace…

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -128,7 +128,7 @@ uint16_t max_display_update_time = 0;
     extern bool powersupply;
   #endif
-
+        
```
Excess whitespace…

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -1920,7 +1920,20 @@ void kill_screen(const char* lcd_msg) {
           STATIC_ITEM(MSG_MOVE_Z, true, true); break;
         default:
-          STATIC_ITEM(MSG_MOVE_E, true, true); break;
+         #if E_MANUAL == 1
```
Indent by 1 more space here.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -1920,7 +1920,20 @@ void kill_screen(const char* lcd_msg) {
+         #if E_MANUAL == 1
+            STATIC_ITEM(MSG_MOVE_E, true, true); break;
+          #else
+            switch(axis - E_AXIS) {
```
One space between `switch` and the opening parenthesis.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -2315,9 +2339,9 @@ void kill_screen(const char* lcd_msg) {
     void _reset_e0_acceleration_rate() { _reset_e_acceleration_rate(0); }
     void _reset_e1_acceleration_rate() { _reset_e_acceleration_rate(1); }
-    #if E_STEPPERS > 2
+    #if EXTRUDERS > 2
```
Again, we have to be very careful here. If this happens to be needed for `SINGLENOZZLE_SWITCHING_EXTRUDER` then what is more likely needed is (perhaps yet another) conditional value. But maybe the `E_MANUAL` value is what you want here and the items below…?

schimmi70 replied:
there is no defenition of E_MANUAL in conditionals_LCD.h ... What is the meaning of the parameter ?

schimmi70 replied:
there are many Labels that don't really fit to current possible configurations. 
The switching extruder have 2 extruder with only 1 stepper but it is can be usefull to define another stepping parameter for the 2 extruder (i have flexible filament, that need more stepps fore the same extrude length, cause it is squeezed by the spring of the extruder)
If you can say me with parameter i should can use, i will change it.

thinkyhead reviewed (changes requested):
Please address these comments.

schimmi70 committed:
requested changes of last commit

schimmi70 committed:
Merge branch 'RCBugFix' into SINGLENOZZLE_SWITCHING_EXTRUDER

# Conflicts:
#	Marlin/Marlin_main.cpp
#	Marlin/SanityCheck.h
#	Marlin/stepper_indirection.h
#	Marlin/ultralcd.cpp

schimmi70 committed:
resolved conflicts in configuration_store.cpp

schimmi70 committed:
resolve conflicts configuration_store.cpp / second try

schimmi70 committed:
third try

schimmi70 committed:
next attempt

schimmi70 committed:
Changes must be to save all 4 parameters, cause there are only 2 Stepper and 4 Extruder...

schimmi70 commented:
changes are made... please tell me what i have to do now !?

schimmi70 committed:
Merge branch 'RCBugFix' into SINGLENOZZLE_SWITCHING_EXTRUDER

# Conflicts:
#	Marlin/configuration_store.cpp

schimmi70 committed:
Changes must be to save all 4 parameters, cause there are only 2 Stepper for 4 Extruder...

thinkyhead commented on Marlin/Conditionals_LCD.h:
```diff
@@ -322,8 +330,8 @@
-  #if ENABLED(DISTINCT_E_FACTORS) && E_STEPPERS > 1
-    #define XYZE_N (XYZ + E_STEPPERS)
+  #if ENABLED(DISTINCT_E_FACTORS) && EXTRUDERS > 1
+    #define XYZE_N (XYZ + EXTRUDERS)
```
This breaks Marlin.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -157,7 +157,19 @@
   #define SWITCHING_EXTRUDER_SERVO_NR 0
   #define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90 } // Angles for E0, E1
-  //#define HOTEND_OFFSET_Z {0.0, 0.0}
+  #define HOTEND_OFFSET_Z {0.0, 0.0}
```
Don't change this in the main configuration.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -157,7 +157,19 @@
+  #define HOTEND_OFFSET_Z {0.0, 0.0}
+#endif
+
+// A single extruder that uses one or two single stepper motor
```
"A single extruder that uses one or two single stepper motor**s**"

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -157,7 +157,19 @@
+#endif
+
+// A single extruder that uses one or two single stepper motor
+// Don't forget to set SSDE_SERVO_ANGLES 
```
"`SSDE_SERVO_ANGLES`" is nowhere to be found.

schimmi70 replied:
at line 155 same mistake ;-)

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -449,7 +449,7 @@
+#define LCD_TIMEOUT_TO_STATUS 60000
 
 #if ENABLED(SDSUPPORT)
 
```
Don't include changes to the main configurations unless they pertain to the feature.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -6403,11 +6403,18 @@ inline void gcode_M77() { print_job_timer.stop(); }
  * M104: Set hot end temperature
  */
+#if ENABLED(SINGLENOZZLE) || ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
+  static int16_t singlenozzle_temp [4];
```
No space between an array variable name and its opening brace.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -6403,11 +6403,18 @@ inline void gcode_M77() { print_job_timer.stop(); }
 
-  #if ENABLED(SINGLENOZZLE)
+  #if ENABLED(SINGLENOZZLE) || ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
+    // save temperatur for different extruder material
```
"temperatur**e**"

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9276,6 +9286,22 @@ inline void gcode_M999() {
+      MOVE_SERVO(SWITCHING_EXTRUDER_SERVO_E1_E2_NR, angles_e1_e2[e]);
+    else      
+      MOVE_SERVO(SWITCHING_EXTRUDER_SERVO_E3_E4_NR, angles_e3_e4[e - 2]);
+  #endif
```
Indent everything here from `#if` to `#endif` by 2 spaces.
http://marlinfw.org/docs/development/coding_standards.html

schimmi70 committed:
conditionals_LCD: DISTINCT_E_FACTORS settings, now use only EXTRUDERS settings if SWITCHING_EXTRUDERS are present
Configuration.h, Configuration_adv.h: set to defaults, correct comments
Marlin_main.cpp: correct formating, correct comments

schimmi70 committed:
correct a merge error

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -160,6 +160,18 @@
+#if ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
+  #define SWITCHING_EXTRUDER_SERVO_E1_E2_NR 0
+  #define SWITCHING_EXTRUDER_SERVO_E1_E2_ANGLES { 0, 90 } // Angles for E1, E2
+  #if EXTRUDERS > 2 
```
`EXTRUDERS > 3` ?

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -2408,9 +2432,9 @@ void kill_screen(const char* lcd_msg) {
+    #if E_STEPPERS > 2 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 2)
       void _reset_e2_acceleration_rate() { _reset_e_acceleration_rate(2); }
-      #if E_STEPPERS > 3
+      #if E_STEPPERS > 3 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 3)
```
One more E stepper has been added since this PR.

thinkyhead commented:
Can this be combined with #6704, I wonder? They are very similar.

thinkyhead commented:
With the latest iteration it looks like #6704 has all special cases covered. A servo can be used to switch the filament drive (extruder) and another one can be used to raise/lower one of the nozzles, and these can exist independently. The only thing missing (perhaps) is automating it so that each set of two extruders would automatically have a servo associated with them. So, perhaps we can patch up #6704 to accomplish that.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -11857,10 +11898,24 @@ void manage_inactivity(bool ignore_stepper_queue/*=false*/) {
+      
+      #if ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER)
+        switch (active_extruder) {
+          case 0,1:
```
`case 0,1:` throws an error in my compiler. Use `case 0: case 1:` instead.

thinkyhead commented on Marlin/Version.h:
```diff
@@ -1,4 +1,4 @@
-/**
+﻿/**
```
Use "utf-8" encoding, and avoid "utf-8 with bom."

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -2408,9 +2432,9 @@ void kill_screen(const char* lcd_msg) {
     void _reset_e0_acceleration_rate() { _reset_e_acceleration_rate(0); }
     void _reset_e1_acceleration_rate() { _reset_e_acceleration_rate(1); }
-    #if E_STEPPERS > 2
+    #if E_STEPPERS > 2 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 2)
```
This change doesn't do anything. `E_STEPPERS > 2` already covers the case. Same for all others below.

schimmi70 replied:
That isn't right... In the case aof singlenozzle switching extruder, you have 1 E_STRPPER but 2 EXTRUDER or 
2 E_STEPPERS and 4 EXTRUDER. So if you want to control all 4 Extruder differently, you must use EXTRUDER not  E_STEPPERS to match al 4 posibble settings (example: differnt material, that use more stepps , like rubber)

schimmi70 commented:
to: #6704  i'm not sure i understand you right. 
Do you mean it can be possible that up to 2 switching extruder are used with a Switching Nozzle or 2 Hotends !?

Must i resolve the conflicts !?

schimmi70 committed:
Update Marlin_main.cpp

requested canges

schimmi70 committed:
Update Version.h

i hoppe now it is OK.

schimmi70 commented:
For 1 1/2 months, I try to make it right for  thinkyhead. Now comes fixoid and 1 week later my changes thrown completely overboard !? I think I'm tired to pull anything here ! Do whatever you want, I am out and use this feature only for Me ... Thank you very much !

Roxy-3D commented:
>  -    #if E_STEPPERS > 2
> +    #if E_STEPPERS > 2 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 2)

Please build a Truth Table for various values and the resulting expression.   I think you will find the second half of the expression doesn't change the results.

schimmi70 commented:
@ Roxy-3D:

One Switching Extruder :  E_STEPPERS = 1  /  EXTRUDERS = 2
Two Switching Extruder :  E_STEPPERS = 2  /  EXTRUDERS = 4

for two switching extruder we have following cases
E_STEPPERS > 2  = false
E_STEPPERS > 2 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 2) = true

E_STEPPERS > 3 = false
E_STEPPERS > 3 || (ENABLED(SINGLENOZZLE_SWITCHING_EXTRUDER) && EXTRUDERS > 3) = true

Now you can say, one stepper motor need only one setting, but if you use same matrial like rubber i need more steps for 1 mm as for ABS Filament .  
Also it confuse if you can use comand T0 - T3 for Extruder but for DISTINCT_E_FACTORS you have only Parameter for T0 and T1.

Roxy-3D commented:
Thank you for the explanation.   I understand things better now...

schimmi70 commented:
to: SWITCHING_EXTRUDER improvement - double switching extruder support #6774

I think the best way is to make it possible, to combinate switching extruder with single_nozzle or 2 Hotends or switching nozzle ... but there are some differences, that need clarify. 

- Temperature of differnt filaments, also at one Hotend. 
- Differnt steps per mm and filament 
- Assignment of extruder to Nozzle 
- and so on

It would be good if we make a differnce between Extruder <-> Hotend.  And between Extruder <-> Esteppers.

schimmi70 commented:
HELLO some one here  ??

schimmi70 commented:
HELLO, HELLO .....
